### DISCUSSION TITLE
Fast function to extract image size from JPEG file

### DISCUSSION THREAD

**Discussion by dgolub** [SixLabors/ImageSharp#1966] [Ideas]
My application currently contains a custom function based on the example at https://stackoverflow.com/questions/552467/how-do-i-reliably-get-an-image-dimensions-in-net-without-loading-the-image to extract the image size from a JPEG files.  It can currently get an image size in <1 ms per file.  I recently migrated from System.Drawing to ImageSharp, which generally went very well.  Based on the documentation, I assumed that `Image.Identify` would do the job and that I'd be able to eliminate the code to read JPEG files manually.  However, when I made that change, the time per image went up to a little over 60 ms, which is prohibitive.  It would be great if ImageSharp provided a comparably fast function for this so that users don't have to implement custom JPEG reading code to get good performance.

**Comment by JimBobSquarePants** [comment]
Something is wrong if it’s taking that long.

**Comment by antonfirsov** [comment]
Platform? Runtime? OS? Other information that may give a lead? This operation should be very fast on x86/x64 + coreclr. One possible reason can be that you are on blazor, xamarin or similar and not using AOT.

**Reply by dgolub** [reply]
.NET 6 on Windows 10 with a 64-bit Intel processor.

**Reply by br3aker** [reply]
Can you provide an example image and code you used?
There's no way it takes 60ms, even full jpeg decoding takes less time, let alone `Identify` call.

**Reply by dgolub** [reply]
Apologies for the delay.  The code to reproduce is fairly straightforward:
```
using System.IO;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using SixLabors.ImageSharp;

namespace TestRepro
{
    [TestClass]
    public class ReproTest
    {
        [TestMethod]
        public void TestIdentifyImage()
        {
            using Stream stream = File.OpenRead(@"D:\TestRepro\695n0001.jpg");
            IImageInfo info = Image.Identify(stream);
            Assert.AreEqual(800, info.Width);
            Assert.AreEqual(600, info.Height);
        }
    }
}
```

The image is available from https://www.eastcoastroads.com/states/ny/inter/i695/fullsize/695n0001.jpg.

**Reply by br3aker** [reply]
How did you benchmark this code? Please don't say 'unit test execution time'...

[Did a very rough benchmarking setup](https://gist.github.com/br3aker/58276881a7cf9dab2a01f1bf1d26c6cf) (BenchmarkDotNet is better) and got these results:

```
System.Drawing.Image
Total time for 1000 iterations: 129ms
Average: 0,129ms

StackOverflow manual binary parsing code
Total time for 1000 iterations: 58ms
Average: 0,058ms

ImageSharp
Total time for 1000 iterations: 331ms
Average: 0,331ms
```

While these results are far beyond from being '60ms per image' it's still slower than `System.Drawing`, right? Well, it's not really true as your image has exif metadata:
![image](https://user-images.githubusercontent.com/20967409/153793352-fbf2a770-e608-460e-a617-1df952496028.png)

Which is automatically parsed by ImageSharp even in `Identify` call. If we would remove that exif parsing code ImageSharp actually surpasses `System.Drawing`:

```
System.Drawing.Image
Total time for 1000 iterations: 132ms
Average: 0,132ms

StackOverflow code
Total time for 1000 iterations: 57ms
Average: 0,057ms

ImageSharp
Total time for 1000 iterations: 115ms
Average: 0,115ms
```

Binary parsing stuff from StackOverflow is still faster because manual binary search for exact bytes would always be faster than layers of abstractions :)

**Reply by antonfirsov** [reply]
We need an `Identify` overload that can disable metadata parsing. This is a tricky API design problem though, since `IgnoreMetadata` is a property of individual format-specific decoder types, and an `Identify` overload taking an `IImageDecoder` feels counter-intuitive. Probably good as a start though. (In the case of the current discussion for example the format is known.)

Sidenote:
@dgolub please don't take it as the lack of such overloads is your problem. As noted in https://github.com/SixLabors/ImageSharp/discussions/1966#discussioncomment-2170248, you should not see 60ms in a proper benchmark.

**Reply by dgolub** [reply]
Thanks for looking into this!  Any idea why the test showed as running for 60 ms when I used `Image.Identify` but dropped to less than 1 ms when I switched it back to the code from Stack Overflow?  Some sort of overhead with calling into a different assembly or something?

**Reply by antonfirsov** [reply]
Could it be related to https://github.com/SixLabors/ImageSharp/issues/1809#issuecomment-965699139?

**Reply by br3aker** [reply]
> Any idea why the test showed as running for 60 ms

.net overhead, JIT, etc.
`System.Drawing` is actually native C++ GDI wrapper for .net so there's almost no overhead, that binary parsing thing has too little code for any significant overhead. ImageSharp has to compile a lot of things before the first launch but it should be a lot faster on second `Identify` call.

**Reply by luoxgwb** [reply]
``` ini

BenchmarkDotNet=v0.13.1, OS=Windows 10.0.22000
Intel Core i7-7700 CPU 3.60GHz (Kaby Lake), 1 CPU, 8 logical and 4 physical cores
.NET SDK=6.0.200
  [Host]     : .NET 5.0.14 (5.0.1422.5710), X64 RyuJIT  [AttachedDebugger]
  DefaultJob : .NET 5.0.14 (5.0.1422.5710), X64 RyuJIT


```
|            Method |     Mean |   Error |  StdDev |      Min |      Max |  Gen 0 | Allocated |
|------------------ |---------:|--------:|--------:|---------:|---------:|-------:|----------:|
| TestIdentifyImage | 207.0 μs | 4.04 μs | 5.11 μs | 201.5 μs | 217.7 μs | 2.1973 |     79 KB |
